wgpu-types/src/lib.rs

```diff
@@ -7243,3 +7245,154 @@ pub enum DeviceLostReason {
     /// will call the callback immediately, with this reason.
     DeviceInvalid = 4,
 }
+
+/// An internal counter for debugging purposes
+///
+/// Internally represented as an atomic isize if the `counters` feature is enabled,
+/// or compiles to nothing otherwise.
+pub struct InternalCounter {
+    #[cfg(feature = "counters")]
+    value: AtomicIsize,
+}
+
+impl InternalCounter {
+    /// Creates a counter with value 0.
+    #[inline]
+    pub const fn new() -> Self {
+        InternalCounter {
+            #[cfg(feature = "counters")]
+            value: AtomicIsize::new(0),
+        }
+    }
+
+    /// Get the counter's value.
+    #[cfg(feature = "counters")]
+    #[inline]
+    pub fn read(&self) -> isize {
+        self.value.load(Ordering::Relaxed)
+    }
+
+    /// Get the counter's value.
+    ///
+    /// Always returns 0 if the `counters` feature is not enabled.
+    #[cfg(not(feature = "counters"))]
+    #[inline]
+    pub fn read(&self) -> isize {
+        0
+    }
+
+    /// Get and reset the counter's value.
+    ///
+    /// Always returns 0 if the `counters` feature is not enabled.
+    #[cfg(feature = "counters")]
+    #[inline]
+    pub fn take(&self) -> isize {
+        self.value.swap(0, Ordering::Relaxed)
+    }
+
+    /// Get and reset the counter's value.
+    ///
+    /// Always returns 0 if the `counters` feature is not enabled.
+    #[cfg(not(feature = "counters"))]
+    #[inline]
+    pub fn take(&self) -> isize {
+        0
+    }
+
+    /// Increment the counter by the provided amount.
+    #[inline]
+    pub fn add(&self, _val: isize) {
+        #[cfg(feature = "counters")]
+        self.value.fetch_add(_val, Ordering::Relaxed);
+    }
+
+    /// Decrement the counter by the provided amount.
+    #[inline]
+    pub fn sub(&self, _val: isize) {
+        #[cfg(feature = "counters")]
+        self.value.fetch_add(-_val, Ordering::Relaxed);
+    }
+
+    /// Sets the counter to the provided value.
+    #[inline]
+    pub fn set(&self, _val: isize) {
+        #[cfg(feature = "counters")]
+        self.value.store(_val, Ordering::Relaxed);
+    }
+}
+
+impl Clone for InternalCounter {
+    fn clone(&self) -> Self {
+        InternalCounter {
+            #[cfg(feature = "counters")]
+            value: AtomicIsize::new(self.read()),
+        }
+    }
+}
+
+impl Default for InternalCounter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl std::fmt::Debug for InternalCounter {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.read().fmt(f)
+    }
+}
+
+/// `wgpu-hal`'s internal counters.
+#[derive(Clone, Default)]
+pub struct HalCounters {
+    // API objects
+    ///
+    pub buffers: InternalCounter,
+    ///
```

Comment:
issue: Newer Clippy versions will reject this (see also #5651); we should either document these here, or bite the bullet and use lint management to indicate intent without a warning:

```suggestion
    #[allow(empty_docs)]
```

Reply:
Implemented my suggestion with a491a5e46.

Reply:
style: It'd be nice to have another newline between the `// section header` comments that are also confusable with the doc. comments for fields.

Implemented this suggestion with 436f5e124; feel free to reject.

Reply:
😮‍💨 Oh, sigh, both of these get reverted by the `rustfmt` we're using as part of `rust-toolchain.toml`. Guess I'll revert these.